src/metabase/api/api_key.clj

```diff
@@ -29,28 +30,84 @@
         (user/set-permissions-groups! user [(perms-group/all-users) group_id])
-        (-> (t2/insert-returning-instance! :model/ApiKey
-                                           {:user_id      (u/the-id user)
-                                            :name         name
-                                            :unhashed_key api-key
-                                            :created_by   api/*current-user-id*})
-            (select-keys [:created_at :updated_at :id])
-            (assoc :name name
-                   :group_id group_id
-                   :unmasked_key api-key
-                   :masked_key (api-key/mask api-key)))))))
+        (let [api-key (-> (t2/insert-returning-instance! :model/ApiKey
+                                                         {:user_id      (u/the-id user)
+                                                          :name         name
+                                                          :unhashed_key unhashed-key
+                                                          :created_by   api/*current-user-id*})
+                          (t2/hydrate :group_name))]
+          (events/publish-event! :event/api-key-create
+                                 {:object  api-key
+                                  :user-id api/*current-user-id*})
+          (-> api-key
+              (select-keys [:created_at :updated_at :id :group_name :name])
+              (assoc :unmasked_key unhashed-key
+                     :masked_key (api-key/mask unhashed-key))))))))
 
 (api/defendpoint GET "/count"
   "Get the count of API keys in the DB"
   [:as _body]
   (api/check-superuser)
   (t2/count :model/ApiKey))
 
+(api/defendpoint PUT "/:id"
+  "Update an API key by changing its group and/or its name"
+  [id :as {{:keys [group_id name] :as _body} :body}]
+  {id       ms/PositiveInt
+   group_id [:maybe ms/PositiveInt]
+   name     [:maybe ms/NonBlankString]}
+  (api/check-superuser)
+  (t2/with-transaction [_conn]
+    (let [api-key-before (-> (t2/select-one :model/ApiKey :id id)
+                             ;; hydrate the group_name for audit logging
+                             (t2/hydrate :group_name))]
+      (when name
+        ;; A bit of a pain to keep these in sync, but oh well.
+        (t2/update! :model/User (:user_id api-key-before) {:first_name name})
+        (t2/update! :model/ApiKey id {:name name}))
+      (when group_id
+        (let [user (-> api-key-before (t2/hydrate :user) :user)]
+          (user/set-permissions-groups! user [(perms-group/all-users) {:id group_id}])))
+      (let [updated-api-key (-> (t2/select-one :model/ApiKey :id id)
+                                (t2/hydrate :group_name))]
+        (events/publish-event! :event/api-key-update {:object updated-api-key
+                                                      :previous-object api-key-before
+                                                      :user-id api/*current-user-id*})
```

Comment:
Sorry about all the code churn here, but I don't think we want to be publishing an event while still in the transaction, since there could be a number of downstream DB calls. It's _probably_ OK but I think it would be safest to limit the transaction to just the calls that need to happen atomically...if the audit log DB write fails for some reason it doesn't need to block the main operation